yt/frontends/chimera/io.py

```diff
@@ -0,0 +1,231 @@
+        # chunks is a list of chunks, and ptf is a dict where the keys are
+        # ptypes and the values are lists of fields.
+        pass
+
+    def _read_particle_fields(self, chunks, ptf, selector):
+        # This gets called after the arrays have been allocated.  It needs to
+        # yield ((ptype, field), data) where data is the masked results of
+        # reading ptype, field and applying the selector to the data read in.
+        # Selector objects have a .select_points(x,y,z) that returns a mask, so
+        # you need to do your masking here.
+        pass
+
+    def _read_fluid_selection(self, chunks, selector, fields, size):
+        rv = {}
+        nodal_fields = []
+        for field in fields:
+            finfo = self.ds.field_info[field]
+            nodal_flag = finfo.nodal_flag
+            if np.any(nodal_flag):
+                num_nodes = 2 ** sum(nodal_flag)
+                rv[field] = np.empty((size, num_nodes), dtype="=f8")
+                nodal_fields.append(field)
+            else:
+                rv[field] = np.empty(size, dtype="=f8")
+        ind = 0
+        for field, mesh, data in self.io_iter(chunks, fields):
+            if data is None:
+                continue
+            else:
+                ind += mesh.select(selector, data.flatten(), rv[field], ind)  # caches
+        return rv
+
+    def io_iter(self, chunks, fields):
+        for n, chunk in enumerate(chunks):
+            file = _find_files(self.filename)
+            with h5py.File(file[n], "r") as f:
+                # Generates mask according to the "ongrid_mask" variable
+                m = int(file[n][-5:-3]) - 1
+                k = f["fluid"]["entropy"].shape[0]
+                mask_0 = f["mesh"]["ongrid_mask"][k * m : k * (m + 1), :]
+
+                if f["mesh"]["array_dimensions"][2] > 1:
+                    nrd = f["mesh"]["array_dimensions"][0] - 2
+                else:
+                    nrd = f["mesh"]["array_dimensions"][0]
+
+                mask = np.repeat(mask_0[:, :, np.newaxis], nrd, axis=2).transpose()
+                for field in fields:  # Reads data by locating subheading
+                    ftype, fname = field
+                    specials = (
+                        "abar",
+                        "e_rms_1",
+                        "e_rms_2",
+                        "e_rms_3",
+                        "e_rms_4",
+                        "lumin_1",
+                        "lumin_2",
+                        "lumin_3",
+                        "lumin_4",
+                        "num_lumin_1",
+                        "num_lumin_2",
+                        "num_lumin_3",
+                        "num_lumin_4",
+                        "shock",
+                        "nse_c",
+                    )
+
+                    a_name_2 = [i.decode("utf-8") for i in f["abundance"]["a_name"]]
+                    a_name_dict = {name.strip(): name for name in a_name_2}
```

Comment:
these definitions do not depend on the iterated variable and could be taken out of the loop